[CORL-1334] Keyboard shortcuts

nick-funk opened the pull request:
## What does this PR do?

Allows you to use the `C` and `Shift + C` key to traverse through the comment stream.

## Changes to schema?

None

## How to test

- Go to All Comments in stream
- Press `C` to traverse through visible comment tree
- Press `Shift + C` to reverse traverse through comments

nick-funk committed:
Create keypress forwarding for Pym based iframe

CORL-1334

nick-funk committed:
Move key registrations to a hook

CORL-1334

nick-funk committed:
Create KeyboardShortcuts component

Will get ready for using relay as our comment tree.

CORL-1334

wyattjoh commented on src/core/client/stream/App/useKeys.ts:
```diff
@@ -0,0 +1,48 @@
+function useKeys() {
+  const [key, setKey] = useState(null);
+
+  function handleKeyMessage(e: any) {
```
This function should be wrapped in a `useCallback`.

wyattjoh commented on src/core/client/stream/common/KeyboardShortcuts/KeyboardShortcuts.tsx:
```diff
@@ -0,0 +1,70 @@
+}
+
+const KeyboardShortcuts: FunctionComponent = () => {
+  const handleKeyMessage = useCallback((e: any) => {
```
Should use concrete types instead of `any`.

nick-funk committed:
Create preliminary keyboard shortcut comment traversal

- Still need to replace X key with Shift+C to go backwards
when walking through comments

CORL-1334

nick-funk committed:
Extract out element to KeyStop transform

CORL-1334

nick-funk committed:
Use `Shift+C` to go back in keyboard comment traversal

CORL-1334

nick-funk committed:
Use pym to scroll during keyboard shortcut events

CORL-1334

nick-funk committed:
Rename keyboard shortcut element traversal for clarity

CORL-1334

nick-funk committed:
Protect the RTE, report dialog from keyboard shortcut events

CORL-1334

nick-funk committed:
Update snapshots

CORL-1334

nick-funk committed:
Remove unused useKeys

CORL-1334

nick-funk committed:
Declare type on key event handlers

CORL-1334

wyattjoh commented on src/core/client/embed/PymControl.ts:
```diff
@@ -15,7 +15,7 @@ export const defaultPymControlFactory: PymControlFactory = (config) =>
 
 export default class PymControl {
-  private pym: pym.Parent;
+  public pym: pym.Parent;
```
Revert to `private` as it doesn't seem it's used elsewhere.

wyattjoh commented on src/core/client/embed/StreamEmbed.ts:
```diff
@@ -167,6 +168,8 @@ export class StreamEmbed {
       url,
     });
+
+    hookUpWindowEvents(this.pymControl);
```
Use the existing `streamDecorators` array above and rewrite the `hookUpWindowEvents` as a decorator so it supports a unsubscribe function.

See https://github.com/coralproject/talk/blob/master/src/core/client/embed/decorators/withClickEvent.ts for an example.

wyattjoh commented on src/core/client/stream/common/KeyboardShortcuts/KeyboardShortcuts.tsx:
```diff
@@ -0,0 +1,191 @@
+  }, []);
+
+  const toKeyStop = useCallback((el: Element) => {
+    const id = el.attributes.getNamedItem("data-keyid");
```
Data attributes can be retrieved instead via:

```
el.dataset.keyid
```

wyattjoh replied:
Note that you must use `HTMLElement` instead of `Element` here to allow this.

wyattjoh replied:
Note that if you rewrite access for data attributes to use kebab case, access via `.dataset` is done via camel case:

```
el.dataset.isLoadMore
```

wyattjoh commented on src/core/client/stream/tabs/Comments/Stream/AllCommentsTab/AllCommentsTabContainer.tsx:
```diff
@@ -264,6 +266,8 @@ export const AllCommentsTabContainer: FunctionComponent<Props> = ({
               aria-controls="comments-allComments-log"
               className={CLASSES.allCommentsTabPane.loadMoreButton}
+              data-keystop={true}
+              data-isLoadMore={true}
```
Data props typically are written in kebab case:

```
data-key-stop
data-is-load-more
```

alinhle replied:
I don't think that there is a convention.
`aria` props does not separate words:
```
aria-labelledby
aria-describedby
```
either way is fine for me though.

wyattjoh committed:
review: addressed review comments

wyattjoh committed:
fix: updated snapshots

kodiakhq[bot] committed:
Merge branch 'master' into feature/CORL-1334-keyboard-shortcuts
